lib/Dialect/HW/Transforms/HWVectorization.cpp

```diff
@@ -0,0 +1,879 @@
+
+  bool all_bits_have_same_source() const;
+  bool is_linear(int size, mlir::Value sourceInput);
+  bool is_reverse(int size, mlir::Value sourceInput);
+
+  mlir::Value getSingleSourceValue() const;
+  size_t size() const;
+};
+
+class vectorizer {
+public:
+  vectorizer(hw::HWModuleOp module);
+
+  hw::HWModuleOp module;
+  llvm::DenseMap<mlir::Value, bit_array> bit_arrays;
+
+  mlir::Value findBitSource(mlir::Value vectorVal, unsigned bitIndex,
+                            int depth = 0);
+  mlir::Value
+  vectorizeSubgraph(OpBuilder &builder, mlir::Value slice0Val,
+                    unsigned vectorWidth,
+                    llvm::DenseMap<mlir::Value, mlir::Value> &vectorizedMap);
+
+  bool can_vectorize_structurally(mlir::Value output);
+  bool areSubgraphsEquivalent(
+      mlir::Value slice0Val, mlir::Value sliceNVal, unsigned sliceIndex,
+      int stride, llvm::DenseMap<mlir::Value, mlir::Value> &slice0ToNMap);
+  bool isValidPermutation(const llvm::SmallVector<unsigned> &perm,
+                          unsigned bitWidth);
+  bool can_apply_partial_vectorization(Value oldOutputVal);
+
+  bool hasCrossBitDependencies(mlir::Value outputVal);
+  void collectLogicCone(mlir::Value val, llvm::DenseSet<mlir::Value> &cone);
+  bool isSafeSharedValue(mlir::Value val,
+                         llvm::SmallPtrSetImpl<mlir::Value> &visited);
+  bool isSafeSharedValue(mlir::Value val);
+
+  void process_extract_ops();
+  void process_concat_ops();
+
+  void process_or_op(comb::OrOp op);
+  void process_and_op(comb::AndOp op);
+  void process_logical_ops();
+  void process_xor_op(comb::XorOp op);
+
+  void vectorize();
+
+  void apply_linear_vectorization(mlir::Value oldOutputVal,
+                                  mlir::Value sourceInput);
+  void apply_reverse_vectorization(mlir::OpBuilder &builder,
+                                   mlir::Value oldOutputVal,
+                                   mlir::Value sourceInput);
+  void apply_mix_vectorization(mlir::OpBuilder &builder,
+                               mlir::Value oldOutputVal,
+                               mlir::Value sourceInput,
+                               const llvm::SmallVector<unsigned> &map);
+  void apply_structural_vectorization(OpBuilder &builder,
+                                      mlir::Value oldOutputVal);
+  void apply_partial_vectorization(OpBuilder &builder,
+                                   mlir::Value oldOutputVal);
+
+  bool cleanup_dead_ops(Block &body);
+};
+
+bit_array::bit_array(llvm::DenseMap<int, bit> &bits) : bits(bits) {}
+bit_array::bit_array(const bit_array &other) : bits(other.bits) {}
+bit_array::bit_array() : bits(llvm::DenseMap<int, bit>()) {}
+
+bool bit_array::all_bits_have_same_source() const {
+  llvm::DenseSet<mlir::Value> sources;
+  for (const auto &[_, bit] : bits) {
+    if (!sources.contains(bit.source))
+      sources.insert(bit.source);
+    if (sources.size() >= 2)
+      return false;
+  }
+  return true;
```

Comment:
super nit: using DenseSet is certainly overkill here, e.g.:
```suggestion
  mlir::Value source;
  for (const auto &[_, bit] : bits) {
    if(source && source != bit.source) return false;
    source = bit.source;
  }
  return true;
```